### ISSUE TITLE
Proposal for BaseRegistryImposition

### ISSUE THREAD

**Issue by danveloper** [ratpack/ratpack#1057]
To start with, I'll acknowledge that this proposal has a very specific, fringe audience, who is interested in building frameworks that sit on top of Ratpack.

***tl;dr: Allow components to be bound before the user registry is evaluated and without relying on Guice, Spring, etc...***

The use case for having an imposition on the base registry is one where a framework (library) may wish to provide a specific set of components to applications via the most basic use of `Registry`, and at a point in the bootstrapping flow that is prior to the user registry being evaluated. That is to say, without relying on the use of Guice or Spring or some other module-oriented DI framework that backs a registry, but also not strictly precluding their use by the user.

Also, an ideal outcome is one where the usage of the overset framework would not differ dramatically from the way that Ratpack works on its own.

By providing default component bindings at a layer _above_ the user registry, such a library would allow the user registry to override those bindings. Furthermore, doing so at the base registry means that the user registry would be able to leverage any registry backing it requires. The underlying framework (that sits on top of Ratpack), could then be completely agnostic to any implementation details made at the user registry and above.

An simple example of this framework-style code may look something like:

```java
package com.company.framework;

import ratpack.server.RatpackServer;

public class Framework {

  public static RatpackServer start(Action<RatpackServerSpec> definition) throws Exception {
    return RatpackServer.of(
                 BaseRegistryImposition.of(r -> r
                   .add(FrameworkDatabaseInterface.class, new PGDatabase())
                   .add(SomeOtherComponent.class, new FrameworkObject())
                 ),
                 definition
    );
  }
}
```

A user could then utilize the `Framework` class to bootstrap their app, in the same way they would otherwise:

```java
package com.company.app;

import static com.comapny.framework.Framework.start;

public class Main {

  public static void main(String[] args) throws Exception {
    start(spec -> spec
      .handlers(chain -> chain
        .all(ctx -> {
          // `db` is a `PGDatabase` instance
          FrameworkDatabaseInterface db = ctx.get(FrameworkDatabaseInterface.class);
          db.doSomeStuff().then(results ->
            ctx.render(results)
          )
        })
      )
    );
  }
}
```

Now at this point, application's that need more specific bindings could handle that via their own user registry:

```java
public class Main {

  public static void main(String[] args) throws Exception {
    start(spec -> spec
      .registry(Guice.registry(b -> b
        .bind(FrameworkDatabaseInterface.class, RedisDatabase.class)
      ))
      .handlers(chain -> chain
        .all(ctx -> {
          // `db` is now a `RedisDatabase`
          FrameworkDatabaseInterface db = ctx.get(FrameworkDatabaseInterface.class);
          db.doSomeStuff().then(results ->
            ctx.render(results)
          )
        })
      )
    );
  }
}
```

This would also, of course, mean that other libraries could provide Guice modules or Spring bindings that could be used for the application's more-specific use case.

Today, we have `UserRegistryImposition`, but for this use case we can't use it because it overrides user registry bindings, instead of allowing them to override the defaults.

Implementing this is trivial, so I guess the proposal here is more of a design discussion than anything. _Should we do this?_

**Comment by johnrengelman** [comment]
I'm not sure impositions is the right thing for this. Those feel more test oriented but I guess it is a method to dealing with it. 
I think I'd prefer a specific API for extending the framework in these scenarios. It could be implemented using something similar to impositions but would be a separate construct. Something like:
ratpack {
  extend(Workflow) {
    //Workflow specific DSL
  }
}

Where Workflow extends some base class that we incorporate into th framework init

**Comment by danveloper** [comment]
@johnrengelman yeah that's ultimately what i'm trying to accomplish with this issue, but thinking more abstractly about the problem, we could provide something that would also be useful to frameworks that are just looking to provide base-level component binding opinions without necessarily extending the DSL.

I think extending the DSL is something separate, but may be able to use the same mechanism for providing base-level component bindings that we spec out here.

**Comment by johnrengelman** [comment]
My worry is that something like `BaseRegistryImposition` will be abused by users.
Though to be fair, the easiest & quickest way to implement this would be through an `Imposition`.

This is kind of touching on the edge of an actual ratpack plugin/module system (beyond registry). Some official way to adding capabilities to the core framework itself.
